std/experimental/allocator/building_blocks/ascending_page_allocator.d

```diff
@@ -0,0 +1,418 @@
+            enforce(data != null, "Failed to VirtualAlloc memory");
+        }
+        offset = data;
+    }
+
+    /**
+    Round the allocation to the next multiple of the page size.
+    The allocation only reserves a range of virtual pages but the actual
+    physical memory is allocated on demand, when accessing the memory
+    Returns `null` on failure or if the requested size exceeds the remaining capacity.
+    */
+    void[] allocate(size_t n)
+    {
+        import std.exception : enforce;
+
+        size_t goodSize = goodAllocSize(n);
+        if (offset - data > numPages * pageSize - goodSize)
+            return null;
+
+        void* result = offset;
+        offset += goodSize;
+        pagesUsed += goodSize / pageSize;
+
+        // Give memory back just by changing page protection
+        version(Posix)
+        {
+            import core.sys.posix.sys.mman : mmap, mprotect,
+                   MAP_PRIVATE, MAP_ANON, MAP_FAILED, PROT_NONE, PROT_WRITE, PROT_READ;
+            auto ret = mprotect(result, goodSize, PROT_WRITE | PROT_READ);
+            enforce(ret == 0, "Failed to allocate memory, mprotect failure");
+        }
+        else version(Windows)
+        {
+            import core.sys.windows.windows : VirtualProtect, PAGE_READWRITE;
+            uint oldProtect;
+            auto ret = VirtualProtect(result, goodSize, PAGE_READWRITE, &oldProtect);
+            enforce(ret != 0, "Failed to allocate memory, VirtualProtect failure");
+        }
+
+        return cast(void[]) result[0 .. n];
+    }
+
+    /**
+    Rounds the requested size to the next multiple of the page size
+    */
+    size_t goodAllocSize(size_t n)
+    {
+        return n.roundUpToMultipleOf(pageSize);
+    }
+
+    /**
+    Removes the read/write protection of the pages which the passed buffer covers.
+    Hints the OS to reclaim the physical memory(``madvise`` for Posix and
+    ``VirtualUnlock`` for Windows).
+    If the allocator is no longer valid, and all objects have been deallocated,
+    the range of virtual addresses is unmapped.
+    */
+    version(Posix)
+    {
+        bool deallocate(void[] buf)
+        {
+            import core.sys.posix.sys.mman : posix_madvise, POSIX_MADV_DONTNEED, mprotect, PROT_NONE, munmap;
+            import std.exception : enforce;
+
+            size_t goodSize = goodAllocSize(buf.length);
+            auto ret = mprotect(buf.ptr, goodSize, PROT_NONE);
+            enforce(ret == 0, "Failed to deallocate memory, mprotect failure");
+
+            // madvise to let the OS reclaim the resources
+            ret = posix_madvise(buf.ptr, goodSize, POSIX_MADV_DONTNEED);
+            enforce(ret == 0, "Failed to deallocate, posix_madvise failure");
+            pagesUsed -= goodSize / pageSize;
+
+            if (!valid && pagesUsed == 0)
+            {
+                munmap(data, numPages * pageSize);
+                data = null;
```

Comment:
I have a feeling you should reset `offset` to null as well. Also it doesn't seem that `allocate` would gracefully fail, except on mprotect failure. Maybe there should be `assert(valid)` there.